File: src/cpp/camera/canny.cc
#include "common/camera.h"

#include <opencv2/imgcodecs.hpp>

int main(int /*argc*/, char const */*argv*/[]) {
  Camera cam(0);
  if (!cam.IsOpened()) {
    std::cerr << "ERROR: Open camera failed" << std::endl;
    return 1;
  }
  std::cout << "\033[1;32mPress ESC/Q to terminate\033[0m\n\n";

  cv::Mat edges;
  cam.Capture([&cam, &edges](const cv::Mat &frame,
      const cv::Mat &/*depthmap*/) {
    cv::cvtColor(frame, edges, cv::COLOR_BGR2GRAY);
    cv::GaussianBlur(edges, edges, cv::Size(7, 7), 1.5, 1.5);
    cv::Canny(edges, edges, 0, 30, 3);
    cam.DrawInfo(edges);
    cv::imshow("edges", edges);

    char key = static_cast<char>(cv::waitKey(10));
    return !(key == 27 || key == 'q' || key == 'Q');  // ESC/Q
  });

  return 0;
}

File: src/python/camera/canny.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# pylint: disable=missing-docstring,import-error,wrong-import-order
import cv2
import numpy as np

from common.camera import Camera


def main():
  def callback(frame, _):
    edges = cv2.Canny(frame, 100, 200)
    cv2.imshow('frame', np.vstack((frame, edges)))

  with Camera(0) as cam:
    print('Camera: %dx%d, %d' % (
        cam.get(cv2.CAP_PROP_FRAME_WIDTH),
        cam.get(cv2.CAP_PROP_FRAME_HEIGHT),
        cam.get(cv2.CAP_PROP_FPS)))
    cam.capture(callback, True)

  cv2.destroyAllWindows()


if __name__ == '__main__':
  main()


# Canny Edge Detection:
#   http://docs.opencv.org/master/da/d22/tutorial_py_canny.html
